File: src/Lessons/4-CountingElements/4.1-FrogRiverOne.go
package main

import "fmt"

func main() {

	fmt.Println(solution(3, []int{1, 3, 1, 3, 1, 2, 1, 3}))
}

func solution(position int, someArray []int) int {
	seconds := -1
	//create array to track steps taken
	//Get the total sum of required steps from 1 to position
	stepsRequired := ((1 + position) * position) / 2 //is a series
	stepsTaken := make([]bool, position)             //create array to store if a step is taken

	for index := 0; index < len(someArray); index++ {
		currentPosition := someArray[index]
		if currentPosition <= position && !stepsTaken[currentPosition-1] {
			stepsRequired -= currentPosition
			stepsTaken[currentPosition-1] = true
		}
		if stepsRequired == 0 {
			seconds = index
			break
		}
	}
	return seconds
}

File: src/Lessons/4-CountingElements/4.1-FrogRiverOne.js
function solution(X, A) {
    //Find how many seconds to get to X in array A;
    let array = [...A];
    let seconds = -1;
    let stepTracker = ((1 + X) * X) / 2;
    let stepsTaken = new Array(X).fill(false); //loop 1;

    //loop 2
    for (let index = 0; index < array.length; index++) {
        const element = array[index];
        if (element <= X && !stepsTaken[element - 1]) {
            stepsTaken[element - 1] = true;
            stepTracker -= element;
        }
        if (stepTracker == 0) {
            seconds = index;
            break;
        }
    }
    return seconds;
}
const testArray = [1, 3, 1, 3, 1, 2, 1, 3];
console.log(solution(3, testArray))
//expect 7
